driver/src/test/functional/com/mongodb/reactivestreams/client/gridfs/GridFSPublisherSpecification.groovy

```diff
@@ -378,43 +408,58 @@ class GridFSPublisherSpecification extends FunctionalSpecification {
     }
 
     def 'should cleanup when unsubscribing'() {
-        when:
-        def contentBytes = multiChunkString as byte[]
+        given:
+        def contentSize = 1024 * 1024
+        def contentBytes = new byte[contentSize]
+        new SecureRandom().nextBytes(contentBytes)
+        def options = new GridFSUploadOptions().chunkSizeBytes(1024)
+        def data = (0..1024).collect { ByteBuffer.wrap(contentBytes) }
+        def subscriber = new Subscriber<ObjectId>() {
+            Subscription subscription
+            @Override
+            void onSubscribe(final Subscription s) {
+                subscription = s
+            }
 
-        then:
-        run(filesCollection.&countDocuments) == 0
+            @Override
+            void onNext(final ObjectId o) {
+            }
+
+            @Override
+            void onError(final Throwable t) {
+            }
+
+            @Override
+            void onComplete() {
+            }
+        }
 
         when:
-        def subscriber = new ObservableSubscriber()
-        gridFSBucket.uploadFromPublisher('myFile', createPublisher(ByteBuffer.wrap(contentBytes), ByteBuffer.wrap(contentBytes)))
+        gridFSBucket.uploadFromPublisher('myFile', createPublisher(*data), options)
             .subscribe(subscriber)
-        subscriber.getSubscription().request(1)
-
-        then:
-        !subscriber.isCompleted()
+        subscriber.subscription.request(1)
 
         then:
+        retry(10) { run(chunksCollection.&countDocuments) > 0 }
         run(filesCollection.&countDocuments) == 0
-        tryMultipleTimes({ run(chunksCollection.&countDocuments) }, 5)
 
-        then:
-        subscriber.getSubscription().cancel()
+        when:
+        subscriber.subscription.cancel()
 
         then:
+        retry(50) { run(chunksCollection.&countDocuments) == 0 }
```

Comment:
Is this racy?  Is there a chance that we could finish writing a file before cancel happens?

Reply:
There is a chance - but with such a large publisher it would have to write 1024 chunks to the db and a file data very quickly.

We use the retry block to test that some data has been written. As soon as that passes we cancel and then retry again 50 times to ensure its cleaned up.

Unfortunately, I couldn't think of a better way to test this.